modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/VFSTransportListener.java

```diff
@@ -1195,4 +1222,74 @@ public int compare(FileObject o1, FileObject o2) {
             return lDiff.intValue();
         }
     }  
+
+    class RequiredFileNamesComparator implements Comparator<FileObject> {
+        String fileNamePattern = null;
+        String requiredFileNamePatterns = null;
+
+        public RequiredFileNamesComparator(String fileNamePattern, String requiredFileNamePatterns){
+            this.fileNamePattern = fileNamePattern;
+            this.requiredFileNamePatterns = requiredFileNamePatterns;
+        }
+
+        @Override
+        public int compare(FileObject o1, FileObject o2) {
+        	//compairs the string of both file names by ignoring the "variable" part
+        	// ex. abc-file1.hl7 and def-file1.pdf with regex [a-z]$1[.]pdf,[a-z]$1[.]hl7 will compair only the "file1" string to sort the files
+        	/**
+        	 * complex example:
+        	 	Regex: [a-z]+[-]$1[-][1-9]+[-]$2[.]txt,[a-z]+[-]$2[-][1-9]+[-]$1[.]hl7
+				INPUT files in folder
+				aaa-testFileA-12345-date1.txt
+				bbb-testFileA-2344-date2.txt
+				ccc-testFileB-45345-date1.txt
+				ddd-date1-32423-testFileA.hl7
+				eee-date1-123-testFileB.hl7 
+				fff-date2-2342-testFileA.hl7
+				
+				SORTED files list
+				aaa-testFileA-12345-date1.txt
+				ddd-date1-32423-testFileA.hl7
+				bbb-testFileA-2344-date2.txt
+				fff-date2-2342-testFileA.hl7
+				ccc-testFileB-45345-date1.txt
+				eee-date1-123-testFileB.hl7 
+        	 
+        	 */
+            String fileName1 = removeFixedPatterns(o1, requiredFileNamePatterns);
+            String fileName2 = removeFixedPatterns(o2, requiredFileNamePatterns);
+            return fileName1.compareTo(fileName2);
+        }
+
+        private String removeFixedPatterns(FileObject child, String requiredFileNamePatterns) {
+        	//get base name of file
+        	String baseName = child.getName().getBaseName();
+        	//separate the patterns
+        	String[] patterns = requiredFileNamePatterns.split(",");
+        	for(int i=0;i<patterns.length;i++){
+        		String fixedPattern = patterns[i];
+        		ArrayList<String> groups = new ArrayList<>();
+        		//extract the $1 to $9 (max 9 groups can be used for now) and replace them with regex named groups
+        		for (int j = 0; j <= 9; j++) {
+        			if ( fixedPattern.matches(".*[$]["+j+"].*")) {
+            			fixedPattern = fixedPattern.replaceAll("[$]["+j+"]", "(?<grp"+j+">.*)");
+            			groups.add("grp"+j);
+        			}
+				}
+        		Pattern pattern = Pattern.compile(fixedPattern);
+        		Matcher matcher = pattern.matcher(baseName);
+        		//match the "fixed" expression with named groups
+        		if(matcher.find()) {
+        			String relevantname = "";
+        			//reconstruct the filename that matters with the regex groups ($1-$9)
+        			for (String groupName: groups) {
+            			relevantname+=matcher.group(groupName);
+					}
```

Comment:
Shall we format the code segments? It seems both tabs and spaces are used. Let's use only spaces and format the code. Please check other places as well.